doc/l2mc/IGMP_Snooping_HLD.md

```diff
@@ -0,0 +1,701 @@
+| 0.4 | 01/24/2020  | Ashok Krishnegowda, Benzeer Bava | Incorporated review comments	   |
+
+
+
+# About this Manual
+This document provides general information about IGMP Snooping feature implementation in SONiC.
+# Scope
+This document describes the functionality and high level design of IGMP Snooping feature in SONiC. 
+
+
+# Definition/Abbreviation
+### Table 1: Abbreviations
+| Term   | Meaning                                                |
+|--------|--------------------------------------------------------|
+| IGMP   | Internet Group Management Protocol            		  |
+| PIM    | Protocol Independent Multicast						  |
+| L2MC	 | Layer-2 Multicast									  |
+|Mrouter | Multicast router										  |
+|MCLAG    | Multi-Chassis Link Aggregation Group                   |
+
+# 1 Feature Overview
+The forwarding of multicast control packets and data through a Layer 2 device configured with VLANs is most easily achieved by the Layer 2 forwarding of received multicast packets on all the member ports of the VLAN interfaces. However, this simple approach is not bandwidth efficient, because only a subset of member ports may be connected to devices interested in receiving those multicast packets. In a worst-case scenario, the data would get forwarded to all port members of a VLAN with a large number of member ports, even if only a single VLAN member is interested in receiving the data. Such scenarios can lead to loss of throughput for a device that gets hit by a high rate of multicast data traffic.
+Internet Group Management Protocol (IGMP) snooping is a mechanism by which a Layer 2 device can effectively address this issue of inefficient multicast forwarding to VLAN port members. Snooping involves "learning" forwarding states for multicast data traffic on VLAN port members from the IGMP control (join/leave) packets received on them.
+When snooping is enabled on a VLAN, switch examines IGMP packets between hosts connected to network switches and multicast routers. Switch builds member list based on what it learns from the report and mrouter port based on IGMP queries or PIM hello packet from multicast router, the switch then forwards multicast traffic to those interfaces in the VLAN where interested receivers are present instead of flooding the traffic to all member ports of the VLAN.
+
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+ 1. Support IGMP Snooping for IGMP Version1 (IGMPv1), IGMPv2, IGMPv3 (INCL Mode).
+ 2. Support IPv4 based forwarding.
+ 3. Support IGMP fast leave.
+ 4. Support IGMP querier functionality.
+ 5. Support static and dynamic multicast router list. 
+ 6. Aging of dynamic mrouter ports will be supported. Static mrouter ports will not timed out, they have to be un-configured explicitly.
+ 7. Support static multicast group configuration. Static l2mc entries will not timed out, they have to be un-configured explicitly.
+ 8. Support trapping IGMP and PIM Hello control packets to CPU.
```

Comment:
is this typo wrt hello packets? why do we need PIM hello packets for IGMP snooping?

Reply:
When device running with IGMP Snooping, it maintains the router ports based on the received PIM hello messages that PIM routers send.